#### Dimentionality Reduction

Use to get maximum information with minimum dimentions (independant variables).

Purpose: To handle less number of variables and do prediction.

One of the Main Technique using for dimentionality reduction is PCA (Principle Component Analysis)

If we have 3 dimensions in x, y, z axes - we will cut into 2 (like we will group into 2) and do predictions with that. 
Problems in dimentionality reduction:
1. The new cuts will not be the actual independant variables.
2. We will lose lot of data.
3. Accuracy may come down.

For example, if we have 5 independant variables in the data. If we apply PCA for dimentionality reduction, it will give 5 different options with the percentage of information covers by each from initial 5 independant variables. We can choose from the new 5 variables to create prediction models based on information percentage those have.

###### Applying dimentionality reduction to dataset

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\statsmodels\tools\_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [6]:
df = pd.read_csv(r'D:\E\Courses\Edureka\Bank-data.csv')

In [7]:
print(df.shape)
df.head(3)

(518, 7)


Index  interest_rate  credit  Gender  previous  duration Churn
0      0          1.334       0       1         0       117    no
1      1          0.767       0       0         1       274   yes
2      2          4.858       0       1         0       167    no

We have 5 independant variables tp predict the churn

In [10]:
# making the churn (categorical) to numeric

df['Churn'] = df['Churn'].astype('category')
df['Churn'] = df['Churn'].cat.codes

In [11]:
df

Index  interest_rate  credit  Gender  previous  duration  Churn
0        0          1.334       0       1         0       117      0
1        1          0.767       0       0         1       274      1
2        2          4.858       0       1         0       167      0
3        3          4.120       0       0         0       686      1
4        4          4.856       0       1         0       159      0
..     ...            ...     ...     ...       ...       ...    ...
513    513          1.334       0       1         0       204      0
514    514          0.861       0       0         1       806      1
515    515          0.879       0       0         0       290      0
516    516          0.877       0       0         1       473      1
517    517          4.965       0       0         0       142      0

[518 rows x 7 columns]

In [12]:
#defining x (independany variables) & y (dependant variables)

x = df.iloc[:, 1:6].values
y = df.iloc[:, 6].values

In [13]:
x

array([[  1.334,   0.   ,   1.   ,   0.   , 117.   ],
       [  0.767,   0.   ,   0.   ,   1.   , 274.   ],
       [  4.858,   0.   ,   1.   ,   0.   , 167.   ],
       ...,
       [  0.879,   0.   ,   0.   ,   0.   , 290.   ],
       [  0.877,   0.   ,   0.   ,   1.   , 473.   ],
       [  4.965,   0.   ,   0.   ,   0.   , 142.   ]])

In [14]:
y

array([0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,

In [15]:
# spliting to train and test data

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2)

In [26]:
from sklearn.linear_model import LogisticRegression

model1 = LogisticRegression()
model1.fit(x_train, y_train)

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [27]:
y_predict = model1.predict(x_test)

In [28]:
from sklearn.metrics import confusion_matrix, accuracy_score

In [29]:
print(confusion_matrix(y_test, y_predict))
accuracy_score(y_test, y_predict)

[[46  8]
 [11 39]]


0.8173076923076923

We are getting 81.7% accuracy score while applying prediction with variables directly.

#### Applying PCA for dimentionality reduction

In [30]:
from sklearn.decomposition import PCA

pca = PCA(n_components=None)

x_train_pca = pca.fit_transform(x_train)
x_test_pca = pca.fit_transform(x_test)

In [32]:
pd.DataFrame(x_train_pca)

# created entirely different components from the original independant variables.

0         1         2         3         4
0    -60.048036 -2.231284  0.554840  0.836036 -0.095575
1    136.951450  2.040389 -0.249206 -0.048789 -0.014967
2   -120.048034 -2.252935  0.558496  0.834306 -0.096320
3     19.951898 -1.943618 -0.224682 -0.329449 -0.054563
4   -203.048169 -1.556343 -0.417187  0.675471 -0.128614
..          ...       ...       ...       ...       ...
409 -108.048143 -1.581839 -0.218685 -0.304443 -0.052203
410  749.951503  2.090148  0.691739  0.164383  0.029276
411 -324.048466  1.279787 -0.218500 -0.103889 -0.026438
412 -103.048160 -1.431659 -0.219644 -0.293861 -0.050705
413  230.951507  2.035279  0.722772  0.158727  0.024117

[414 rows x 5 columns]

In [33]:
# checking the percentage of information each new component (created through PCA) holding.

pca.explained_variance_ratio_

array([9.99963262e-01, 3.39664489e-05, 1.86991836e-06, 6.51649648e-07,
       2.49822234e-07])

Means 1st component is holding 99.996% of the information

In [34]:
# re-applying PCA with 1 component as the 1st component holds 99.996% of the info

pca = PCA(n_components=1)

x_train_pca = pca.fit_transform(x_train)
x_test_pca = pca.fit_transform(x_test)

In [36]:
pd.DataFrame(x_train_pca)

0
0    -60.048036
1    136.951450
2   -120.048034
3     19.951898
4   -203.048169
..          ...
409 -108.048143
410  749.951503
411 -324.048466
412 -103.048160
413  230.951507

[414 rows x 1 columns]

In [37]:
model2 = LogisticRegression()
model2.fit(x_train_pca, y_train)

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [41]:
y_predict2 = model2.predict(x_test_pca)

In [42]:
confusion_matrix(y_test, y_predict2)

array([[41, 13],
       [18, 32]], dtype=int64)

In [43]:
accuracy_score(y_test, y_predict2)

0.7019230769230769

We reduced 5 independant variables to 1 with Principal Component Analysis (PCA) technique in dimentionality reduction. But accuracy score came down to 70.1% from 81.7%.

##### Applying PCA with 5 components to check accuract score with 100% of the information

In [44]:
pca = PCA(n_components=5)

x_train_pca = pca.fit_transform(x_train)
x_test_pca = pca.fit_transform(x_test)

In [45]:
model3 = LogisticRegression()
model3.fit(x_train_pca, y_train)

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [47]:
y_predict_3 = model3.predict(x_test_pca)

In [48]:
confusion_matrix(y_test, y_predict_3)

array([[23, 31],
       [25, 25]], dtype=int64)

In [49]:
accuracy_score(y_test, y_predict_3)

0.46153846153846156

Accuracy score further reduced to 46.1% with all 5 components generated through PCA with 100% of the informationholding inside.

#### Dimemntionalty reduction is not a good technique to apply aways.

#### We can use techniques like reducing 2 or more variables to one if they depends each other to create a new one.
eg: if height and weight are there as 2 independant variables, we can create BMI (body mass index = weight/height) with these 2 and reduce one variable.